pkg/proxy/iptables/proxier.go

```diff
@@ -1042,20 +1046,30 @@ func (proxier *Proxier) syncProxyRules() {
 			proxier.natRules.Write(args)
 		}
 
+		// These chains represent the sets of endpoints to use when internal or
+		// external traffic policy is "Cluster" vs "Local".
 		policyClusterChain := svcInfo.policyClusterChainName
 		policyLocalChain := svcInfo.policyLocalChainName
-		svcXlbChain := svcInfo.xlbChainName
-
-		internalTrafficChain := policyClusterChain
-		externalTrafficChain := policyClusterChain
 
+		// These chains designate which policy chain to use for internal- and
+		// external-destination traffic.
+		internalPolicyChain := policyClusterChain
+		externalPolicyChain := policyClusterChain
 		if svcInfo.NodeLocalInternal() {
-			internalTrafficChain = policyLocalChain
+			internalPolicyChain = policyLocalChain
 		}
 		if svcInfo.NodeLocalExternal() {
-			externalTrafficChain = svcXlbChain
+			externalPolicyChain = policyLocalChain
 		}
 
+		// These chains are where *ALL* rules which match traffic that is
+		// service-destined should jump.  ClusterIP traffic is considered
+		// "internal" while NodePort, LoadBalancer, and ExternalIPs traffic is
+		// considered "external".
+		internalTrafficChain := internalPolicyChain
+		externalTrafficChain := svcInfo.externalChainName // eventually jumps to externalPolicyChain
```

Comment:
Having both `*policyChain` and `*trafficChain`, where `internalTrafficChain` is always the same as `internalPolicyChain`, but `externalTrafficChain` is not always the same as `externalPolicyChain` seems confusing.

I _think_ it would be clearer to have only 3 variables, corresponding to the things you call `internalPolicyChain`, `externalTrafficChain`, and `externalPolicyChain` here. But not with those exact names... It think the variable pointing to the EXT chain needs a name that makes it clearer that it's doing setup / preprocessing / special-casing. (And then I don't care much whether the other two have "Policy" or "Traffic" in their name.)

Reply:
I initially changed `*TrafficChain` to `*DestinationChain` so it was clearer that all of the `-j` code jumped to either `internalDestination` or `externalDestination` - but it was a lot more diff for less obvious value.

The extra level of indirection was intended to make clear that `policy{Cluster,Local}Chain` means "the chain to use for this policy" and `{internal,external}PolicyChain` means "which of cluster/local to use for internal/external", and `{internal,external}TrafficChain` was the indirection to hide the details of where to start.

Concretely, I though about adding an INT to match EXT, but INT is *always* empty.  

I also considered `policy{Cluster,Local}Chain` + `{itp,etp}Chain` + `{internal,external}DestChain`.  Or we could put this into a function like `internalDestChain, externalDestChain := chooseDestChains(svcInfo)`

I will look at this again.  If you're directionally OK with this PR, that's my main concern.

Reply:
I added a commit to add an INT but I don't think I like it perf-wise (but the symmetry is nice).

Reply:
I don't think there's any reason to add an INT chain. My complaint was just that the naming was confusing, not that the organization of the chains was bad.

> I initially changed `*TrafficChain`

yeah, I think "Traffic" is the problem. Like, `externalPolicyChain` implements external policy (and `externalTrafficChain` does not). That makes sense. But both chains receive external traffic, so what exactly is it that `externalTrafficChain` does that `externalPolicyChain` doesn't?

That's why I was suggesting it needs a name like `externalTrafficSetupChain` or something... But then since there's no setup for internal traffic you don't need a corresponding extra variable for internal traffic.